### PR TITLE
Fix global config issue with script command

### PR DESCRIPTION
**Pull request by samj1912**
This should the last of such bugs. Went through the codebase and ensured we were not using global config elsewhere like this.

### CODE REVIEW

File: src/hera/shared/_global_config.py [hunk 1]
```diff
@@ -27,12 +29,12 @@ class _GlobalConfig:
...
     _image: Union[str, Callable[[], str]] = "python:3.7"
     service_account_name: Optional[str] = None
-    script_command: Optional[List[str]] = ["python"]
+    script_command: Optional[List[str]] = field(default_factory=lambda: ["python"])
     _pre_build_hooks: Optional[_HookMap] = None
 
     def reset(self) -> None:
         """Resets the global config container to its initial state"""
-        self.__dict__.clear()  # Wipe instance values to fallback to the class defaults
+        self.__dict__ = _GlobalConfig().__dict__
```
Comment (samj1912) [inline src/hera/shared/_global_config.py:37]: Creating a new instance and sourcing values from there to ensure it sets it to defaults.

File: pyproject.toml [hunk 2]
```diff
@@ -48,6 +48,8 @@ types-PyYAML = "*"
 jsonpath-ng = "^1.5.3"
 datamodel-code-generator = {extras = ["http"], version = "^0.17.1"}
 types-requests = "^2.28.11.12"
+pytest-clarity = "^1.0.1"
+pytest-sugar = "^0.9.6"
```
Comment (samj1912) [inline pyproject.toml:52]: Added these to make it easier to comprehend pytest diff outputs.

File: src/hera/shared/_global_config.py [hunk 3]
```diff
@@ -9,6 +10,7 @@
 _HookMap = Dict[Type[TBase], List[Hook]]
 
 
+@dataclass
```
Comment (flaviuvadan) [inline src/hera/shared/_global_config.py:13]: What do you think about making this a Pydantic model?

Comment (samj1912) [reply src/hera/shared/_global_config.py:13]: @flaviuvadan tried pydantic first, but we cannot mimic the current implementation using pydantic in a b/w compatible fashion.

Pydantic gives errors like - 

```
FAILED tests/test_hooks.py::test_container_pre_build_hooks - ValueError: "_GlobalConfig" object has no field "_pre_build_hooks"
FAILED tests/test_hooks.py::test_workflow_pre_build_hooks - ValueError: "_GlobalConfig" object has no field "_pre_build_hooks"
FAILED tests/test_workflow.py::test_hera_output[coinflip] - ValueError: "_GlobalConfig" object has no field "__dict__"
FAILED tests/test_workflow.py::test_hera_output[complex_deps] - ValueError: "_GlobalConfig" object has no field "__dict__"
FAILED tests/test_workflow.py::test_hera_output[dag_conditional_parameters] - ValueError: "_GlobalConfig" object has no field "__dict__"
FAILED tests/test_workflow.py::test_hera_output[dag_diamond_with_callable_container] - ValueError: "_GlobalConfig" object has no field "__dict__"
FAILED tests/test_workflow.py::test_hera_output[dag_diamond_with_callable_decorators] - ValueError: "_GlobalConfig" object has no field "__dict__"
```

File: examples/workflows/global-config.yaml [hunk 4]
```diff
@@ -11,12 +11,17 @@ spec:
...
+  - name: say
     script:
-      command: ['python']
-      image: 'image-say'
-      source: |
-        import os
+      command:
+      - python3
+      image: image-say
+      source: 'import os
```
Comment (flaviuvadan) [inline examples/workflows/global-config.yaml:19]: Is there an extra space in the script str generation step? Perhaps we could remove it?

**Review by flaviuvadan — APPROVED** [review]
LGTM! Just had a question about script parsing newlines

**Pull request merged**